rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+        }
+
+        self.request_buffer.extend(input);
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.request_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("request: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({} => '{:?}')", current.len(), rem.len(), rem);
+                    current = rem;
+                    self.handle_msg(msg, false);
+                }
+                Err(nom::Err::Incomplete(v)) => {
+                    if let nom::Needed::Size(s) = v {
+                        SCLogDebug!("incomplete request: needed {}", s);
+                    }
+                    self.request_buffer.extend_from_slice(current);
+                    break;
+                }
+                Err(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> bool {
+        if input.len() == 0 {
+            return true;
+        }
+
+        self.response_buffer.extend(input);
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.response_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({})", current.len(), rem.len());
+                    current = rem;
+                    self.handle_msg(msg, true);
+                }
+                Err(nom::Err::Incomplete(v)) => {
+                    if let nom::Needed::Size(s) = v {
+                        SCLogDebug!("incomplete response: needed {}", s);
+                    }
+                    self.response_buffer.extend_from_slice(current);
+                    break;
+                }
+                Err(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn set_event(tx: &mut MQTTTransaction, event: MQTTEvent) {
+        let ev = event as u8;
+        core::sc_app_layer_decoder_events_set_event_raw(&mut tx.events, ev);
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&MQTTTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
```

Comment:
Why minus one ?

Reply:
This is 1:1 from the applayer template code, I didn;t change it. I guess this has to do with the fact that transaction ID 0 is never given out.

Reply:
@jasonish thoughts here ?

Reply:
Yeah, @satta is correct. The ID in the parser is always off by one from the ID outside the parser for reasons that predate me.

Reply:
Thanks Jason. Let me know if you understand the reason someday